### ISSUE TITLE
Conflicts of internal Teuchos with that from Trilinos library

### ISSUE THREAD

**Issue by jppelteret** [symengine/symengine#1516]
I've been developing some features in an external library that uses both SymEngine and Trilinos. Because I've been adding features to SymEngine as well I've built it in debug mode, while my fully-featured Trilinos is built in release mode. This causes some conflicts because SymEngine builds Teuchos with `HAVE_TEUCHOS_STACKTRACE` defined, while in Trilinos `HAVE_TEUCHOS_STACKTRACE` is not defined. Depending on the order in which the Trilinos and SymEngine headers are defined this conflict results in the following error:
```
<path>/install-dev/include/symengine/symengine_rcp.h:279:16: error: ‘Teuchos::print_stack_on_segfault’ has not been declared
 using Teuchos::print_stack_on_segfault;
```
The issue goes away if I build SymEngine in release mode (and I guess that it would if I built Trilinos in debug mode, thereby keeping these both consistent). I think that the real solution here might be to prefix/suffice all of the `#defined` macros in the bundled Teuchos, thereby allowing these two separate implementation in the separate libraries to coexist in a conflict-free manner.

**Comment by certik** [comment]
@jppelteret good point. Thanks for the tip. I wrote the stacktrace facility for Trilinos originally. Another solution is to make it a small separate library and make both Trilinos and SymEngine depend on it, but that's long term. For now we should prefix the macros.
